build: also support KUBE_RACE for test binaries

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This is relevant for building e2e.test in
pull-kubernetes-e2e-kind-alpha-beta-features-race: if it contains data races, tests might be flaky.

#### Which issue(s) this PR is related to:

Follow-up to: https://github.com/kubernetes/kubernetes/pull/133834

#### Special notes for your reviewer:

Contains one fix for the prior PR (indention by two spaces).

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @dims

pohly committed:
build: also support KUBE_RACE for test binaries

This is relevant for building e2e.test in
pull-kubernetes-e2e-kind-alpha-beta-features-race: if it contains data races,
tests might be flaky.

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

aojea commented:
IIUIC and you build the kubernetes binaries with the race detector then those are the ones that will panic on races, i.e. kube-apiserver, kubelet, kube-controller-manager , ... so it is not only that ginkgo will not report them, is that the cluster will be unstable and components will be restarting

pohly commented:
> is that the cluster will be unstable and components will be restarting ... 

No, the race detector doesn't kill the binary. It just prints the reports and then continues.

> and most probably not even the job will be able to finish

It worked fine over in https://github.com/kubernetes/kubernetes/pull/133844, both in the sense that the E2E run finished without any failures in the tests themselves and in the sense that it found new data races in system components.

What didn't work here is running e2e.test with race detection. I suspect the OOM killer gets triggered. I'll get back to this later.

aojea commented:
> No, the race detector doesn't kill the binary. It just prints the reports and then continues.

Today I learned ... thanks

pohly committed:
E2E: treat data races in e2e suite as failures

Ginkgo itself doesn't do this, in which case prun-junit-xml drops the output
and Spyglass wouldn't show the test as failed. If the data race warning is
captured, we now treat that as the failure of a test if it hasn't already
failed for other reasons.

While at it, the entire report cleanup gets moved to our junit package.

pohly committed:
test: automatically lower Ginkgo parallelism when using race detection

When race detection is enabled, merely running 25 e2e.test instances was too much
and the OOM killer shut down the Prow test pod because of the memory overhead.

A CI job could control that via GINKGO_PARALLEL_NODES, but we should also have
saner defaults which take this into account.

pohly commented:
Yeah, data races in JUnit :-)

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133870/pull-kubernetes-e2e-kind-alpha-beta-features-race/1967952266430255104

One is a false positive, a test logs output from a system component which has a DATA RACE. The test itself is fine.

pohly commented:
This is now ready for review.

We could enable race detection in e2e.test without also enabling race detection in the cluster under test, but I think having one dedicated job which enables it in both is better.

pohly commented:
/sig testing

pohly commented:
There's one known issue at the moment: https://github.com/kubernetes/kubernetes/issues/134418

pohly commented:
/assign @aojea

aojea commented on test/e2e/framework/internal/junit/junit.go:
```diff
@@ -42,5 +47,60 @@ func WriteJUnitReport(report ginkgo.Report, filename string) error {
+	// Both variants of captured output get checked, just to be on the safe side.
+	for i, spec := range report.SpecReports {
+		if spec.State == types.SpecStatePassed {
+			ginkgoWriterRace := containsDataRace(spec.CapturedGinkgoWriterOutput)
+			stdoutRace := containsDataRace(spec.CapturedStdOutErr)
```
who writes this output? as , who is responsible of a test that passes make it fail because there is a race?
The test written in ginkgo?

aojea commented:
one question about the junit parsing to make test fails,

aojea commented on test/e2e/framework/internal/junit/junit.go:
```diff
@@ -42,5 +47,60 @@ func WriteJUnitReport(report ginkgo.Report, filename string) error {
+	// the test as failed.
+	//
+	// Both variants of captured output get checked, just to be on the safe side.
+	for i, spec := range report.SpecReports {
```
this seems something that it will run for all jobs once it enabled, can we detect if the golang race detector is enabled and only check this if that is the case?

pohly replied:
Changed.

pohly committed:
E2E framework: enable data race detection only if needed

When building the test binary without race detection, we don't
need the post-processing of the JUnit file because it cannot
contain data race reports. This can be done via build tags.

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race
